toolkit/tools/internal/safechroot/safechroot.go

```diff
@@ -441,50 +451,80 @@ func cleanupAllChroots() {
 			logger.Log.Infof("Cleaning up chroot (%s)", activeChroots[i].rootDir)
-			err := activeChroots[i].unmountAndRemove(leaveChrootOnDisk)
+			err := activeChroots[i].unmountAndRemove(leaveChrootOnDisk, unmountTypeLazy)
 			// Perform best effort cleanup: unmount as many chroots as possible,
 			// even if one fails.
 			if err != nil {
 				logger.Log.Errorf("Failed to unmount chroot (%s)", activeChroots[i].rootDir)
+				failedToUnmount = true
 			}
 		}
 	}
 
-	logger.Log.Info("Cleanup finished")
+	if failedToUnmount {
+		logger.Log.Fatalf("Failed to unmount a chroot, manual unmount required. See above errors for details on which mounts failed.")
+	} else {
+		logger.Log.Info("Cleanup finished")
+	}
 }
 
 // unmountAndRemove retries to unmount directories that were mounted into
 // the chroot until the unmounts succeed or too many failed attempts.
 // This is to avoid leaving folders like /dev mounted when the chroot folder is forcefully deleted in cleanup.
 // Iff all mounts were successfully unmounted, the chroot's root directory will be removed if requested.
-func (c *Chroot) unmountAndRemove(leaveOnDisk bool) (err error) {
+// If doLazyUnmount is true, use the lazy unmount flag which will allow the unmount to succeed even if the mount point is busy.
+func (c *Chroot) unmountAndRemove(leaveOnDisk, lazyUnmount bool) (err error) {
 	const (
-		totalAttempts = 3
-		retryDuration = time.Second
-		unmountFlags  = 0
+		retryDuration      = time.Second
+		totalAttempts      = 3
+		unmountFlagsNormal = 0
+		// Do a lazy unmount as a fallback. This will allow the unmount to succeed even if the mount point is busy.
+		// This is to avoid leaving folders like /dev mounted if the chroot folder is forcefully deleted by the user. Even
+		// if the mount is busy at least it will be detached from the filesystem and will not damage the host.
+		unmountFlagsLazy = unix.MNT_DETACH
+		errMsg           = "Failed to unmount (%s). Error: %s"
 	)
+	unmountFlags := unmountFlagsNormal
+	if lazyUnmount {
+		logger.Log.Warnf("Final attempt to unmount chroot (%s), using a lazy unmount", c.rootDir)
+		unmountFlags = unmountFlagsLazy
+	}
 
 	for _, mountPoint := range c.mountPoints {
 		fullPath := filepath.Join(c.rootDir, mountPoint.target)
 
+		isMounted := true
+		isMounted, err = mountinfo.Mounted(fullPath)
+		if err != nil {
+			err = fmt.Errorf("failed to check if mount point (%s) is mounted. Error: %s", fullPath, err)
+			return err
+		}
+		if !isMounted {
+			logger.Log.Debugf("Skipping unmount of (%s) because it is not mounted", fullPath)
+			continue
+		}
+
 		logger.Log.Debugf("Unmounting (%s)", fullPath)
 
 		// Skip mount points if they were not successfully created
 		if !mountPoint.isMounted {
 			continue
 		}
 
-		err = retry.Run(func() error {
-			return unix.Unmount(fullPath, unmountFlags)
-		}, totalAttempts, retryDuration)
+		_, err = retry.RunWithExpBackoff(func() error {
+			logger.Log.Debugf("Calling unmount (%s, %v)", fullPath, unmountFlags)
```

Comment:
Could you mention in the logs what `%s` and `%v` is? The path might be clear, but the flags might be harder to guess.